src/Database/LSMTree/Internal/Run.hs

```diff
@@ -30,4 +35,127 @@
 -- The above list is a sketch. Functionality may move around, and the list is
 -- not exhaustive.
 --
-module Database.LSMTree.Internal.Run () where
+module Database.LSMTree.Internal.Run (
+    RunLocation (..)
+  , runKOpsPath
+  , runFilterPath
+  , runIndexPath
+  , runBlobPath
+
+  , Run
+  , fromWriteBuffer
+  ) where
+
+import qualified Control.Monad.ST as ST
+import qualified Data.ByteString.Builder as BS
+import qualified Data.ByteString.Short as BS
+import           Data.Foldable (for_)
+import           Database.LSMTree.Common (SomeSerialisationConstraint (..))
+import           Database.LSMTree.Internal.Entry
+import           Database.LSMTree.Internal.Run.BloomFilter (Bloom)
+import           Database.LSMTree.Internal.Run.Construction (BlobRef, RawValue)
+import qualified Database.LSMTree.Internal.Run.Construction as Cons
+import           Database.LSMTree.Internal.Run.Index.Compact (CompactIndex)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey)
+import qualified Database.LSMTree.Internal.Serialise as S
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import           System.FilePath ((</>))
+import qualified System.IO as IO
+
+-- TODO: makes sense? in any case, it should live somewhere else
+data RunLocation = RunLocation
+    { runSessionRoot :: !FilePath
+    , runNumber      :: !Int
+    }
+
+runKOpsPath :: RunLocation -> FilePath
+runKOpsPath = runFilePathWithExt ".keyops"
+
+runFilterPath :: RunLocation -> FilePath
+runFilterPath = runFilePathWithExt ".filter"
+
+runIndexPath :: RunLocation -> FilePath
+runIndexPath = runFilePathWithExt ".index"
+
+runBlobPath :: RunLocation -> FilePath
+runBlobPath = runFilePathWithExt ".blobs"
+
+runFilePathWithExt :: String -> RunLocation -> FilePath
+runFilePathWithExt ext RunLocation {..} =
+    runSessionRoot </> "active" </> (show runNumber <> ext)
+
+-- Long term we want to work with FsPath
+-- runFilePathWithExt :: Strict.Text -> RunLocation -> FsPath
+-- runFilePathWithExt ext RunLocation{..} =
+--   fsPathFromList $
+--     fsPathToList runSessionRoot <> ["active", Strict.pack (show runNumber) <> ext]
+
+--------------------------------------------------------------------------------
+
+-- TODO: what do we want the representation to be like?
```

Comment:
Yes this is ok for now. I have a slightly refined version that looks a lot like this, which I can add as a follow-on PR. (My version has a reference count too, but I forgot the `RunLocation` so that's a useful reminder :-) ).